presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoinsUtils.java

```diff
@@ -0,0 +1,158 @@
+            sets.addAll(generatePartitions(existingSet, totalNodes, index + 1));
+        }
+        else {
+            sets.add(existingSet);
+        }
+        return sets.build();
+    }
+
+    public static JoinNode createBinaryJoin(JoinGraphNode joinGraph, Set<Integer> leftSources, PlanNodeIdAllocator idAllocator)
+    {
+        List<PlanNode> sources = joinGraph.getSources();
+        ImmutableList.Builder<PlanNode> leftNodes = ImmutableList.builder();
+        List<Symbol> leftSymbols = new ArrayList<>();
+        ImmutableList.Builder<PlanNode> rightNodes = ImmutableList.builder();
+        List<Symbol> rightSymbols = new ArrayList<>();
+        for (int i = 0; i < sources.size(); i++) {
+            PlanNode source = sources.get(i);
+            if (leftSources.contains(i)) {
+                leftNodes.add(source);
+                leftSymbols.addAll(source.getOutputSymbols());
+            }
+            else {
+                rightNodes.add(source);
+                rightSymbols.addAll(source.getOutputSymbols());
+            }
+        }
+
+        ImmutableList.Builder<JoinNode.EquiJoinClause> leftCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> rightCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> spanningCriteria = ImmutableList.builder();
+        for (JoinNode.EquiJoinClause equiJoinClause : joinGraph.getCriteria()) {
+            Set<Symbol> equiJoinClauseSymbols = ImmutableSet.of(equiJoinClause.getLeft(), equiJoinClause.getRight());
+            if (leftSymbols.containsAll(equiJoinClauseSymbols)) {
+                leftCriteria.add(equiJoinClause);
+            }
+            else if (rightSymbols.containsAll(equiJoinClauseSymbols)) {
+                rightCriteria.add(equiJoinClause);
+            }
+            else {
+                spanningCriteria.add(equiJoinClause);
+            }
+        }
+
+        ImmutableList.Builder<Expression> leftFilters = ImmutableList.builder();
+        ImmutableList.Builder<Expression> rightFilters = ImmutableList.builder();
+        ImmutableList.Builder<Expression> spanningFilters = ImmutableList.builder();
+        for (Expression filter : joinGraph.getFilters()) {
+            Set<Symbol> filterSymbols = SymbolReferenceExtractor.getSymbolReferences(filter).stream()
+                    .map(Symbol::from)
+                    .collect(toImmutableSet());
+            if (leftSymbols.containsAll(filterSymbols)) {
+                leftFilters.add(filter);
+            }
+            else if (rightSymbols.containsAll(filterSymbols)) {
+                rightFilters.add(filter);
+            }
+            else {
+                spanningFilters.add(filter);
+            }
+        }
+
+        PlanNode left = new JoinGraphNode(idAllocator.getNextId(), leftNodes.build(), leftCriteria.build(), leftFilters.build());
+        if (left.getSources().size() == 1) {
```

Comment:
Maybe instead of creating a `JoinGraphNode` when `leftNodes` contains a single element, we can just do `PlanNode left = leftNode.build().get(0)`. It seems like the `extractOnlySource` method is only used here and I think that method is a little awkward and brittle since it relies on the caller knowing to call it on an instance that has a single source.